src/linking/linkModule.ml

```diff
@@ -599,7 +599,6 @@ let fill_table_base_import new_base : module_' -> module_' = fun m ->
 (* Concatenation of modules *)
 
 let join_modules (em1 : extended_module) (m2 : module_') (ns2 : name_section) : extended_module =
-  assert (m2.start = None);
```

Comment:
So this is really hacky. I think as first thing we should clarify here (in comments/documentation) that this linker and `join_modules` in particular cannot link/join two arbitrary Wasm modules, we have various assumptions like the second module being generated by LLVM and first one by moc (and details related to LLVM- and moc- generated code).

In this case, module 2 is the RTS module and it has a `start` after updating to LLVM 12: `(start $__wasm_apply_global_relocs)`. We add `(call $__wasm_apply_global_relocs)` to the final module after joining modules, so that's why we ignore `start` here.

So it's not OK to ignore `start` in general, we only handle the case specific to LLVM 12 here. If we want this function to link *arbitrary* Wasm modules then this is wrong.

Reply:
But sound it would be cleaner to not ignore the start here, bit rather make sure it is called from the linked start function? But that's what the other comment would yield, I think.

The linker does rely on internals of drun output but ideally not of LLVM internals (like function names in the debug section).

Reply:
Second module's start is now prepended to the merged module's start, so marking this as resolved as well.

Reply:
Sigh.. GitHub UI does not show new inline comments without refreshing the page, so I missed your second comment above.

I guess we could join starts here, instead of ignoring it here and adding module 2's start after `join_modules`. Will update.

Reply:
Done. Not sure if it's any better though..